### PR TITLE
EOS-19810: avoidance of internal dix layout index lookups

### PR DESCRIPTION
**Pull request by imvenkip**
Added idx_pool_version_get() internal API to get index pool version
and layout type. It will be called when m0_entity_create() API is
triggered from consumers like s3, m0kv. It is applicable for create
index operation.

Added required changes to support pver and layout type which are
passed by S3. It is applicable for PUT, GET, and DEL KV operations
in md-path.

Signed-off-by: Venkateswarlu Payidimarry <venkateswarlu.payidimarry@seagate.com>

### CODE REVIEW

File: motr/idx.c [hunk 1]
```diff
@@ -167,6 +226,13 @@ static int idx_op_init(struct m0_idx *idx, int opcode,
 	} else
 		oi->oi_dtx = NULL;
 
+	if (opcode == M0_EO_CREATE) {
+		rc = get_idx_pool_version(oi);
```
Comment (madhavemuri) [inline motr/idx.c:230]: Rename it to idx_pool_version_get()

File: motr/client.h [hunk 2]
```diff
@@ -767,8 +767,16 @@ struct m0_client_layout {
  * - non-distributed index, which is identified by well-formed motr fid having
  *   m0_cas_index_fid_type type.
  */
+
+/* Index attributes */
+struct m0_idx_attr {
+	uint32_t      idx_layout_type;
```
Comment (madhavemuri) [inline motr/client.h:773]: Add comment about attributes.

File: pool/pool.c [hunk 3]
```diff
@@ -669,6 +669,40 @@ m0_pool_version_get(struct m0_pools_common  *pc,
...
+		return M0_RC(-EINVAL);
+
+	m0_tl_for(pools, &pc->pc_pools, pool) {
+		if (is_dix_pool(pc, pool)) {
+			*pv = m0_pool_clean_pver_find(pool);
+			if (*pv != NULL)
+				return M0_RC(0);
+		}
+	} m0_tl_endfor;
+	return M0_RC(0);
```
Comment (madhavemuri) [inline pool/pool.c:690]: Return -ENOENT from here.

File: motr/idx.c (hunk 1 above)
Comment (imvenkip) [reply motr/idx.c:230]: Done

File: pool/pool.c [hunk 4]
```diff
@@ -669,6 +669,40 @@ m0_pool_version_get(struct m0_pools_common  *pc,
 	return M0_RC(rc);
 }
 
+static int dix_pool_version_get_locked(struct m0_pools_common  *pc,
+                                       struct m0_pool_version **pv)
+{
+	struct m0_pool         *pool;
+
+	M0_ENTRY();
+	M0_PRE(m0_mutex_is_locked(&pc->pc_mutex));
```
[the remark this answers is not in the document]
Comment (sg-shankar) [reply pool/pool.c:678]: Thanks for details. This becomes low priority but something we can optimize in future.

**Review by sg-shankar — APPROVED** [review]

File: motr/client.h [hunk 5]
```diff
@@ -767,8 +767,19 @@ struct m0_client_layout {
  * - non-distributed index, which is identified by well-formed motr fid having
  *   m0_cas_index_fid_type type.
  */
+
+/* Index attributes 
+ * idx_layout_type - DIX pool layout type
+ * idx_ver - DIX pool verion
+ */
```
Comment (huanghua78) [inline motr/client.h:774]: Doxygen style comments, please.
```
/**
 * Index attributes.
 * idx_layout_type - DIX pool layout type
 * idx_ver - DIX pool verion
 */
```

File: motr/client_internal.h [hunk 6]
```diff
@@ -942,6 +944,15 @@ M0_INTERNAL void m0__io_ref_put(struct m0_client *m0c);
 M0_INTERNAL struct m0_file *m0_client_fop_to_file(struct m0_fop *fop);
 M0_INTERNAL bool entity_id_is_valid(const struct m0_uint128 *id);
 
+/**
+ * Returns the m0_ client instance, found from the provided index.
```
Comment (huanghua78) [inline motr/client_internal.h:948]:
```
Returns the m0_client instance
```

File: motr/idx.c [hunk 7]
```diff
@@ -93,6 +93,65 @@ M0_INTERNAL bool m0__idx_op_invariant(struct m0_op_idx *oi)
...
+{
+	int               rc;
+	struct m0_client *cinst;
+
+	M0_ENTRY();
+
+	cinst = m0__idx_instance(idx);
+
+	if (pv == NULL)
+		return M0_ERR(-ENOENT);
```
Comment (huanghua78) [inline motr/idx.c:118]: I would like to return -EINVAL to say that invalid parameter is passed by caller.

Comment (huanghua78) [reply motr/idx.c:118]: And this checking can be done at first before calling "cinst = m0__idx_instance(idx);"
Or maybe we can assert that is not NULL.

File: motr/idx.c [hunk 8]
```diff
@@ -93,6 +93,65 @@ M0_INTERNAL bool m0__idx_op_invariant(struct m0_op_idx *oi)
...
+	M0_PRE(M0_IN(OP_IDX2CODE(oi), (M0_EO_CREATE)));
+
+	idx = m0__idx_entity(oi->oi_oc.oc_op.op_entity);
+
+	rc = m0__idx_pool_version_get(idx, &pv);
+	if (rc != 0)
+		return M0_ERR(rc);
+	idx->in_attr.idx_pver = pv->pv_id;
+
+	return M0_RC(rc);
```
Comment (huanghua78) [inline motr/idx.c:152]: we can:
return M0_RC(0);
because we know that rc is 0.

File: motr/idx.c [hunk 9]
```diff
@@ -93,6 +93,65 @@ M0_INTERNAL bool m0__idx_op_invariant(struct m0_op_idx *oi)
...
+	if (rc != 0)
+		return M0_ERR(rc);
+	return M0_RC(rc);
+}
+
+/**
+ * Gets an index DIX pool version. It is applicable for create
+ * index operation.
+ *
+ * @param oi index operation.
```
Comment (huanghua78) [inline motr/idx.c:130]: It's better to add:
```
@param[in][out] oi index operation.
```
to explain that this `oi` will be modified.

File: motr/idx.c [hunk 10]
```diff
@@ -167,6 +226,13 @@ static int idx_op_init(struct m0_idx *idx, int opcode,
 	} else
 		oi->oi_dtx = NULL;
 
+	if ((opcode == M0_EO_CREATE) && (entity->en_type == M0_ET_IDX)) {
+		rc = idx_pool_version_get(oi);
+		if (rc == 0)
+			idx->in_attr.idx_layout_type = DIX_LTYPE_DESCR;
+		return M0_RC(rc);
```
Comment (huanghua78) [inline motr/idx.c:233]:
```
if (rc != 0)
        return M0_ERR(rc);
idx->in_attr.idx_layout_type = DIX_LTYPE_DESCR;
```

File: motr/idx_dix.c [hunk 11]
```diff
@@ -547,8 +547,20 @@ static void cas_next_ast(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 static void dix_build(const struct m0_op_idx *oi,
 		      struct m0_dix          *out)
 {
+	unsigned int   opcode = oi->oi_oc.oc_op.op_code;
```
Comment (huanghua78) [inline motr/idx_dix.c:550]: We have defined:
```
#define OP_OBJ2CODE(op_obj) op_obj->oo_oc.oc_op.op_code

#define OP_IDX2CODE(op_idx) op_idx->oi_oc.oc_op.op_code
```
which can be used here.

If we don't use them, no need to define them.

File: pool/pool.c [hunk 12]
```diff
@@ -669,6 +669,40 @@ m0_pool_version_get(struct m0_pools_common  *pc,
...
+
+	if (pv == NULL)
+		return M0_RC(-EINVAL);
```
Comment (huanghua78) [inline pool/pool.c:681]: return M0_ERR(-EINVAL);

File: pool/pool.c [hunk 13]
```diff
@@ -669,6 +669,40 @@ m0_pool_version_get(struct m0_pools_common  *pc,
...
+
+	m0_tl_for(pools, &pc->pc_pools, pool) {
+		if (is_dix_pool(pc, pool)) {
+			*pv = m0_pool_clean_pver_find(pool);
+			if (*pv != NULL)
+				return M0_RC(0);
+		}
+	} m0_tl_endfor;
+	return M0_RC(-ENOENT);
```
Comment (huanghua78) [inline pool/pool.c:690]: return M0_ERR(-ENOENT);

File: motr/idx.c [hunk 14]
```diff
@@ -93,6 +93,65 @@ M0_INTERNAL bool m0__idx_op_invariant(struct m0_op_idx *oi)
...
+
+	cinst = m0__idx_instance(idx);
+
+	if (pv == NULL)
+		return M0_ERR(-ENOENT);
+
+	rc = m0_dix_pool_version_get(&cinst->m0c_pools_common, pv);
+	if (rc != 0)
+		return M0_ERR(rc);
+	return M0_RC(rc);
```
[the remark this answers is not in the document]
Comment (imvenkip) [reply motr/idx.c:123]: I believe, Return hardcoded values is not good practice.

File: motr/idx.c (hunk 8 above)
Comment (imvenkip) [reply motr/idx.c:152]: I believe, Return hardcoded values is not good practice.

File: motr/idx_dix.c (hunk 11 above)
Comment (imvenkip) [reply motr/idx_dix.c:550]: OP_IDX2CODE(op_idx) -- Already it is used in motr/idx.c file. I will assign opcode value with  OP_IDX2CODE() macro.

File: motr/idx.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (huanghua78) [reply motr/idx.c:123]: If the return is already checked and is known value, it is better to return that known value.
So, the reader of this code don't have to think about the what the value is.
It is more obvious that a success is returned, all error cases are handled already.

**Comment by imvenkip** [comment]
It is created to tmain branch and now tmain is obsolete. 
Created new PR to main branch.  https://github.com/Seagate/cortx-motr/pull/715.

**Pull request closed without merging**